zero/gyrokinetic_maxwellian_correct.c

```diff
@@ -0,0 +1,298 @@
+        nc, moms_target, up->moms_iter, up->abs_diff_moms);
+      gkyl_array_reduce_range(up->error_cu, up->abs_diff_moms, GKYL_MAX, conf_range);
+      gkyl_cu_memcpy(up->error, up->error_cu, sizeof(double[3]), GKYL_CU_MEMCPY_D2H);
+    }
+    else {
+      struct gkyl_range_iter biter;
+
+      // Reset the maximum error
+      up->error[0] = 0.0;
+      up->error[1] = 0.0;
+      up->error[2] = 0.0;
+      // Iterate over the input configuration-space range to find the maximum error
+      gkyl_range_iter_init(&biter, conf_range);
+      while (gkyl_range_iter_next(&biter)){
+        long midx = gkyl_range_idx(conf_range, biter.idx);
+        const double *moms_local = gkyl_array_cfetch(up->moms_iter, midx);
+        const double *moms_target_local = gkyl_array_cfetch(moms_target, midx);
+        // Check the error in the absolute value of the cell average
+        // Note: for density and temperature, this error is a relative error compared to the target moment value
+        // so that we can converge to the correct target moments in SI units and minimize finite precision issues.
+        up->error[0] = fmax(fabs(moms_local[0*nc] - moms_target_local[0*nc])/moms_target_local[0*nc],fabs(up->error[0]));
+        up->error[2] = fmax(fabs(moms_local[2*nc] - moms_target_local[2*nc])/moms_target_local[2*nc],fabs(up->error[2]));
+        // However, u_par may be ~ 0 and if it is, we need to use absolute error. We can converge safely using
+        // absolute error if u_par ~ O(1). Otherwise, we use relative error for u_par. 
+        if (fabs(moms_target_local[1*nc]) < 1.0) {
+          up->error[1] = fmax(fabs(moms_local[1*nc] - moms_target_local[1*nc]),fabs(up->error[1]));
+        }
+        else {
+          up->error[1] = fmax(fabs(moms_local[1*nc] - moms_target_local[1*nc])/moms_target_local[1*nc],fabs(up->error[1]));
+        }
+      }
+    }
+  }
+
+  return (struct gkyl_gyrokinetic_maxwellian_correct_status) {
+    .iter_converged = corr_status,
+    .num_iter = niter,
+    .error[0] = up->error[0], 
+    .error[1] = up->error[1],
+    .error[2] = up->error[2], 
+  };  
+}
+
+void 
+gkyl_gyrokinetic_maxwellian_correct_release(gkyl_gyrokinetic_maxwellian_correct *up)
+{
+  gkyl_array_release(up->moms_iter);
+  gkyl_array_release(up->d_moms);
+  gkyl_array_release(up->dd_moms);
+  gkyl_array_release(up->num_ratio);
+  gkyl_dg_bin_op_mem_release(up->mem);
+  if (up->use_gpu) {
+    gkyl_array_release(up->abs_diff_moms);
+    gkyl_cu_free(up->error_cu);
+  }
+
+  gkyl_gyrokinetic_maxwellian_moments_release(up->moments_up);
+  gkyl_proj_maxwellian_on_basis_release(up->proj_max_prim);
+
+  gkyl_free(up);
+}
+
+#ifndef GKYL_HAVE_CUDA
+
+void 
+gkyl_gyrokinetic_maxwellian_correct_all_moments_abs_diff_cu(const struct gkyl_range *conf_range, 
+  int nc, const struct gkyl_array *moms_target, const struct gkyl_array *moms_iter, 
+  struct gkyl_array *moms_abs_diff)
+{
+  assert(false);
+}
+
+#endif
```

Comment:
@JunoRavin if you move the declaration from the public to the private header I mentioned above, you can just delete this. We should do this in future code and maybe go back and change this in old code because it's cleaner.